### PR TITLE
Fix training resumption

### PR DESCRIPTION
**Pull request by sichu2023**
## Summary
Loss curve from training resumption is inconsistent with a single uninterrupted loss curve.

## Details
It has now been identified partly from incorrect datamodule behavior and is fixed with implementing `state_dict` for datamodule. However, there is still a very minor dip in validation loss curve that does not affect subsequent training curve.

### CODE REVIEW

**Review by jstjohn — APPROVED** [review]

File: sub-packages/bionemo-llm/src/bionemo/llm/data/datamodule.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        Args:
+            dataset (torch.utils.data.Dataset): Dataset to be fed into DataLoader.
+            mode (str): Mode of the DataLoader to enable correct training resumption with datamodule state dict. Must be one of "train", "valid", or "test".
+            **kwargs (Any): Keyword arguments to be passed to torch.utils.data.DataLoader.
+        """
+        super().__init__(dataset, **kwargs)
+
+        if mode not in ["train", "valid", "test"]:
+            raise ValueError(f"Mode must be one of 'train', 'valid', or 'test', but got {mode}")
+        self.mode = mode
```
[the remark this answers is not in the document]
Comment (sichu2023) [reply sub-packages/bionemo-llm/src/bionemo/llm/data/datamodule.py]: We are now sourcing `WrappedDataLoader` from nemo directly.

**Comment by sichu2023** [comment]
Confirm that we can reproduce loss curve upon resumption despite a small jump on validation loss at the beginning.
https://wandb.ai/clara-discovery/bionemo2-esm2-debug/reports/val_ppl-24-10-25-14-39-37---Vmlldzo5ODg0ODMy

**Comment by sichu2023** [comment]
Broke down the PR into smaller pieces. Now we have a PR specifically tackles inconsistency stop-and-go v.s. continuous training curve.
https://github.com/NVIDIA/bionemo-framework/pull/362

**Comment by pstjohn** [comment]
Closing in favor of #375
